subprojects/ide/src/main/java/org/gradle/plugins/ide/eclipse/model/ResourceFilterMatcher.java

```diff
@@ -0,0 +1,143 @@
+/*
+ * Copyright 2011 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.gradle.plugins.ide.eclipse.model;
+
+import com.google.common.base.Objects;
+import com.google.common.collect.Sets;
+import groovy.lang.Closure;
+import groovy.lang.DelegatesTo;
+import org.gradle.api.Action;
+import org.gradle.api.InvalidUserDataException;
+import org.gradle.api.Nullable;
+import org.gradle.api.internal.ClosureBackedAction;
+
+import java.util.Set;
+
+/**
+ * The model of an Eclipse resource filter matcher.
+ * <p>
+ * The matcher decides when the containing filter (or containing matcher) applies.  The
+ * matcher configures things like whether this ResourceFilter matches resources by
+ * name, project relative path, location, last modified, etc.  Eclipse has many types
+ * of built-in matchers and it is possible to specify the id and arguments for custom
+ * matchers using this model.
+ * <p>
+ * A matcher must have an id.  It may have either a custom string argument or a set of nested
+ * child matchers (e.g. an 'or' matcher will have several nested condition matchers).
+ * <p>
+ * For more documentation on usage with examples, see {@link ResourceFilter}.
+ *
+ * @since 3.5
+ */
+public final class ResourceFilterMatcher {
+    private String id;
+    private String arguments;
+    private Set<ResourceFilterMatcher> children = Sets.newLinkedHashSet();
+
+    public ResourceFilterMatcher() {
+    }
+
+    public ResourceFilterMatcher(String id, String arguments, Set<ResourceFilterMatcher> children) {
+        this();
+        this.id = id;
+        this.arguments = arguments;
+        this.children = children;
+    }
+
+    public String getId() {
+        return id;
+    }
+
+    public void setId(String id) {
+        this.id = id;
+    }
+
+    @Nullable
+    public String getArguments() {
+        return arguments;
+    }
+
+    public void setArguments(String arguments) {
+        this.arguments = arguments;
+    }
+
+    public Set<ResourceFilterMatcher> getChildren() {
+        return children;
+    }
+
+    public void setChildren(Set<ResourceFilterMatcher> children) {
```

Comment:
Hi @big-guy , are you asking why nested `ResourceFilterMatcher`s are supported?

This is required by the existing Eclipse model of filters/matchers.  Nested matchers are actually a pretty common use case.  Suppose you wanted to filter out either of two folders from showing in Eclipse, this would be modeled as an `or` matcher with nested matchers for the folder names:

```
eclipse {
  project {
    resourceFilter {
      matcher {
        id = 'org.eclipse.ui.ide.orFilterMatcher'
        matcher {
          id = 'org.eclipse.ui.ide.multiFilter'
          arguments = '1.0-name-matches-false-false-node_modules'
        }
        matcher {
          id = 'org.eclipse.ui.ide.multiFilter'
          arguments = '1.0-name-matches-false-false-build'
        }
      }
    }
  }
}
```

Regarding the DSL issue you mention where `matcher` at the filter level creates/configures a single matcher whereas `matcher` at the matcher level creates a new nested matcher, that is something I struggled with as well.  

I took a look at other examples in the gradle code where the DSL configures single objects vs. collections of objects, and this seemed to be how it was done elsewhere (`EclipseProject.linkedResource`, for example, creates a new Link and adds to the set whereas `EclipseClasspath.file` configures the single `XmlFileContentMerger`).  

The confusion here just seems to stem from the required domain complexity that a filter can have exactly one matcher but matchers themselves can have multiple nested matchers.  I guess the method names could be differentiated (`matcher` vs `nestedMatcher`?), but that might also cause another confusion as to why the same object type is constructed with different keywords depending on its placement in the DSL tree.  Let me know if you have any suggestions.

Please keep in mind that the current implementation is just the "raw" domain model that gets gradle to parity with the Eclipse resource filters, but @oehme [has spec'd a much more intuitive extension to the DSL](https://github.com/gradle/gradle/tree/master/design-docs/features/ide-integration/richer-eclipse-model/resource-filters) that can be layered on top of this.